akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,409 @@
+        system.deadLetters.toImpl.sendSystem(signal)
+        false
+      } else {
+        unsafe.compareAndSwapObject(this, systemQueueOffset, currentList.head, (signal :: currentList).head) || {
+          signal.unlink()
+          needToActivate()
+        }
+      }
+    }
+    try {
+      if (needToActivate()) {
+        val old = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (isClosed(old)) {
+          // nothing to do
+          if (Debug) println(s"[$thread] NOT enqueueing $signal to $self: terminating")
+        } else if (activations(old) == 0) {
+          // all is good: we signaled the transition to active
+          if (Debug) println(s"[$thread] enqueueing $signal to $self: activating")
+          executionContext.execute(this)
+        } else {
+          // take back that token: we didn’t actually enqueue a normal message and the actor was already active
+          if (Debug) println(s"[$thread] enqueueing $signal to $self: already active")
+          unsafe.getAndAddInt(this, statusOffset, -1)
+        }
+      } else if (Debug) println(s"[$thread] NOT enqueueing $signal to $self: terminated")
+    } catch handleException
+  }
+
+  override final def run(): Unit = {
+    if (Debug) println(s"[$thread] entering run($self): interrupted=${Thread.interrupted()}")
+    val status = _status
+    val msgs = messageCount(status)
+    var processed = 0
+    try {
+      unscheduleReceiveTimeout()
+      if (!isClosed(status)) {
+        while (processAllSystemMessages() && !queue.isEmpty() && processed < msgs) {
+          val msg = queue.poll()
+          processed += 1
+          processMessage(msg)
+        }
+      }
+      scheduleReceiveTimeout()
+    } catch {
+      case NonFatal(ex) ⇒ fail(ex)
+      case ie: InterruptedException ⇒
+        fail(ie)
+        if (Debug) println(s"[$thread] interrupting due to catching InterruptedException")
+        Thread.currentThread.interrupt()
+    } finally {
+      // also remove the general activation token
+      processed += 1
+      val prev = unsafe.getAndAddInt(this, statusOffset, -processed)
+      val now = prev - processed
+      if (isClosed(now)) {
+        // we’re finished
+      } else if (activations(now) > 0) {
+        // normal messages pending: reverse the deactivation
+        unsafe.getAndAddInt(this, statusOffset, 1)
+        executionContext.execute(this)
+      } else if (_systemQueue.head != null) {
+        /*
+         * system message was enqueued after our last processing, we now need to
+         * race against the other party because the enqueue might have happened
+         * before the deactivation (above) and hence not scheduled
+         */
+        val again = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (activations(now) == 0) executionContext.execute(this)
+        else unsafe.getAndAddInt(this, statusOffset, -1)
+      }
+    }
+    if (Debug) println(s"[$thread] exiting run($self): interrupted=${Thread.interrupted()}")
+  }
```

Comment:
interesting, it would be great with a description what this important method does and any design reasons